cmrf_views/src/Plugin/views/query/API4.php

```diff
@@ -0,0 +1,525 @@
+   * single DatabaseCondition object, like this:
+   *
+   * @code
+   * $this->query->addWhere(
+   *   $this->options['group'],
+   *   (new Condition('OR'))
+   *     ->condition($field, $value, 'NOT IN')
+   *     ->condition($field, $value, 'IS NULL')
+   * );
+   * @endcode
+   *
+   * @param $group
+   *   The WHERE group to add these to; groups are used to create AND/OR
+   *   sections. Groups cannot be nested. Use 0 as the default group.
+   *   If the group does not yet exist it will be created as an AND group.
+   * @param $field
+   *   The name of the field to check.
+   * @param $value
+   *   The value to test the field against. In most cases, this is a scalar. For more
+   *   complex options, it is an array. The meaning of each element in the array is
+   *   dependent on the $operator.
+   * @param $operator
+   *   The comparison operator, such as =, <, or >=. It also accepts more
+   *   complex options such as IN, LIKE, LIKE BINARY, or BETWEEN. Defaults to =.
+   *   If $field is a string you have to use 'formula' here.
+   *
+   * @see \Drupal\Core\Database\Query\ConditionInterface::condition()
+   * @see \Drupal\Core\Database\Query\Condition
+   */
+  public function addWhere($group, $field, $value = NULL, $operator = NULL) {
+    // Ensure all variants of 0 are actually 0. Thus '', 0 and NULL are all
+    // the default group.
+    if (empty($group)) {
+      $group = 0;
+    }
+
+    // Check for a group.
+    if (!isset($this->where[$group])) {
+      $this->setWhereGroup('AND', $group);
+    }
+
+    $this->where[$group]['conditions'][] = [
+      'field'    => $field,
+      'value'    => $value,
+      'operator' => $operator,
+    ];
+  }
+
+  /**
+   * Generates a unique placeholder used in the API query.
+   */
+  public function placeholder($base = 'views') {
+    static $placeholders = [];
+    if (!isset($placeholders[$base])) {
+      $placeholders[$base] = 0;
+      return ':' . $base;
+    }
+    else {
+      return ':' . $base . ++$placeholders[$base];
+    }
+  }
+
+  /**
+   * Add a complex WHERE clause to the query.
+   *
+   * The caller is responsible for ensuring that all fields are fully qualified
+   * (TABLE.FIELD) and that the table already exists in the query.
+   * Internally the dbtng method "where" is used.
+   *
+   * @param $group
+   *   The WHERE group to add these to; groups are used to create AND/OR
+   *   sections. Groups cannot be nested. Use 0 as the default group.
+   *   If the group does not yet exist it will be created as an AND group.
+   * @param $snippet
+   *   The snippet to check. This can be either a column or
+   *   a complex expression like "UPPER(table.field) = 'value'"
+   * @param $args
+   *   An associative array of arguments.
+   *
+   * @see QueryConditionInterface::where()
+   */
+  public function addWhereExpression($group, $snippet, $args = []) {
+    // Ensure all variants of 0 are actually 0. Thus '', 0 and NULL are all
```

Comment:
Couldn't this replaced by a simple call of `$this->addWhere($group, $snippet, $args, 'formular')`?

Reply:
That's code copied from the APIv3 implementation, let's keep it like that for now.